crates/biome_js_analyze/src/analyzers/nursery/no_nested_test_suites.rs

```diff
@@ -0,0 +1,165 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, AddVisitor, Phases, QueryMatch, Queryable, Rule,
+    RuleDiagnostic, RuleSource, RuleSourceKind, ServiceBag, Visitor, VisitorContext,
+};
+use biome_console::markup;
+use biome_js_syntax::{AnyJsExpression, JsCallExpression, JsLanguage};
+use biome_rowan::{AstNode, Language, SyntaxNode, TextRange, WalkEvent};
+
+declare_rule! {
+    /// This rule enforces a maximum depth to nested `describe()` calls to improve code clarity in your tests,
+    /// with the default option of { "max": 5 }.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// describe('foo', () => {
+    ///   describe('bar', () => {
+    ///     describe('baz', () => {
+    ///       describe('qux', () => {
+    ///         describe('quxx', () => {
+    ///           describe('too many', () => {
+    ///             it('should get something', () => {
+    ///               expect(getSomething()).toBe('Something');
+    ///             });
+    ///           });
+    ///         });
+    ///       });
+    ///     });
+    ///   });
+    /// });
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// describe('foo', () => {
+    ///   describe('bar', () => {
+    ///     it('should get something', () => {
+    ///       expect(getSomething()).toBe('Something');
+    ///   });
+    /// });
+
+    ///   describe('qux', () => {
+    ///     it('should get something', () => {
+    ///       expect(getSomething()).toBe('Something');
+    ///     });
+    ///   });
+    /// });
+    /// ```
+    ///
+    pub NoNestedTestSuites {
+        version: "next",
+        name: "noNestedTestSuites",
+        recommended: false,
+        source: RuleSource::EslintJest("max-nested-describe"),
+        source_kind: RuleSourceKind::Inspired,
+    }
+}
+
+#[derive(Default)]
+struct NestedTestVisitor {
+    stack: Vec<AnyJsExpression>,
```

Comment:
Here are some suggestions, let me know what you think.

Instead of a `Vec`, we can use `SmallVec` instead, because we already know the stack size, which is 5. This means that we will have to implement `Default` manually, using the `with_capacity` function. This will be useful when you'll implement the option.

Then, we can use the function [spilled](https://docs.rs/smallvec/latest/smallvec/struct.SmallVec.html#method.spilled), to check if the vector started to push things in the heap.

Instead of matching all the `describe` that violate the rule, we could just stop at the first one that we find. All the nested describe are already violating the rule.

Reply:
> Then, we can use the function [spilled](https://docs.rs/smallvec/latest/smallvec/struct.SmallVec.html#method.spilled), to check if the vector started to push things in the heap.

Just wondering: Can you pass the future rule option to the visitor to set the vec capacity?

Reply:
Not via `ctx.options`, but a visitor can access the **analyzer options**. It will have to read the options manually, but it's still possible.

Reply:
We don't even to maintain a vec of nodes, just a counter with `curr_val` and `max_val` is fine.